ironfish/src/network/peerNetwork.ts

```diff
@@ -359,17 +372,54 @@ export class PeerNetwork {
     }
   }
 
-  private broadcastTransaction(
-    message: NewTransactionMessage,
-    peersToSendTo: ReadonlyArray<Peer>,
-  ): void {
-    for (const peer of peersToSendTo) {
+  private broadcastTransaction(message: NewTransactionMessage): void {
+    const hash = new Transaction(message.transaction).hash()
+    const isUpgraded = (peer: Peer) => peer.version !== null && peer.version >= 17
+
+    const peersToSendToArray = [...this.connectedPeersWithoutTransaction(hash)]
+    const sendHash: Peer[] = []
+    const sendFull: Peer[] = peersToSendToArray.filter((p) => !isUpgraded(p))
+
+    const upgradedPeers: Peer[] = ArrayUtils.shuffle(
+      peersToSendToArray.filter((p) => isUpgraded(p)),
+    )
+
+    const sqrtSize = Math.floor(Math.sqrt(peersToSendToArray.length))
+
+    for (const peer of upgradedPeers) {
+      if (sendFull.length < sqrtSize) {
+        sendFull.push(peer)
+      } else {
+        sendHash.push(peer)
+      }
+    }
+
+    const hashMessage = new NewPooledTransactionHashes([hash])
+
+    for (const peer of sendHash) {
       if (peer.state.type !== 'CONNECTED') {
         continue
       }
 
-      if (peer.send(message)) {
-        const hash = new Transaction(message.transaction).hash()
+      if (peer.send(hashMessage)) {
+        this.markKnowsTransaction(hash, peer.state.identity)
+      }
+    }
+
+    const newTransactionMessage = new NewTransactionV2Message(message.transaction)
+    for (const peer of sendFull) {
+      if (peer.state.type !== 'CONNECTED') {
+        continue
+      }
+
+      let sent
+      if (isUpgraded(peer)) {
+        sent = peer.send(newTransactionMessage)
+      } else {
+        sent = peer.send(message)
+      }
+
+      if (sent) {
```

Comment:
I think we could simplify this down a bit and make it more like the hash-sending code by doing this:

```suggestion
      const messageToSend = isUpgraded(peer) ? newTransactionMessage : message
      if (peer.send(messageToSend)) {
```

Reply:
Nice yea I like it